### DISCUSSION TITLE
What's the specification of `XpubDerivable`?

### DISCUSSION THREAD

**Discussion by acgxv** [BP-WG/bp-std#8] [Q&A] [answered]
Hi, I'm currently learning this library, but there are no documents or code comments available.

I would like to know how `XpubDerivable` str has been defined because it doesn't seem like a common one.

This is the format that I have frequently encountered:
```
[ccf488ef/m/84'/0'/0']xpub6DA9Gjn12MVZqNV96tL7dZDzzbg6BxYH4JJQmPJRKHASamJEe1X9NSKW8vgjhsevrP3qVbed2BjQcE4Qfk54bgsxUjMerN7iUTQxkXREwiu/*
```

Could you please explain how this is generated?
https://github.com/BP-WG/bp-std/blob/a2005978df1078c3ea55c11db27693c9cd5a9052/std/src/xpub.rs#L624

E.G. Why
- no `m`
- use `h` to replace `'`
- there is a `<0;1>` instead of a single number

Perhaps this question may seem silly, but as a newcomer, and there isn't too much material that I can check. I hope you don't mind me submitting this as an issue.

**Comment by dr-orlovsky** [comment, marked as answer]
Thank you for your interest to the library!

Yes, you are right - docs are not there yet since we are looking into the stabilization of the library (once at least one Bitcoin wallet is released with it) and before that time things may change.

So, your question has two aspects: what is `XpubDerivable` and rationale for formatting its string representation.

`XpubDerivable` is a structure which contains all necessary information to derive a final public key which will be used by a wallet descriptor for constructing scriptPubkeys and PSBTs. Specifically, it consists of two main parts:
- Information about xpub and its origin (`XpubSpec`)
- Information about derivation path from xpub used to derive the final public key (we name it "terminal derivation path")

Xpub origin is required to produce valid PSBT files, thus, we do not have an anonymous `m` and instead provide this origin information as a `[...]` prefix - a standard among Bitcoin wallets today, also used in bitcoin descriptors.

Terminal derivation provides information about the used keychain (a term from BIP32). Frequently it is simply `/*/*`, interpreted by wallets as `/<0;1>/*`, but this is not reliable: with the introduction of RGB and other client-side applications there might be more keychain indexes than just 0 (external) and 1 (internal/chainge). The `/<NUM;NUM;...;NUM>/*` format was introduced in BIP389 and is used here as a better way to address that ambiguity.

Finally, you are free to use both `h` and `'` (both are standard and can be used in an interchanged manner, according to BIP32); to print out `XpubDerivabe` with `'` just format it with alternative format specifier: `format!("{:#}", xpub_derivable)`. We use `h` by default since it is more readable and easier to use in the context of command-line arguments, where `'` will be interpreted as a quoted string.
